silverback/persistence.py

```diff
@@ -0,0 +1,262 @@
+import pickle
+from abc import ABC, abstractmethod
+from datetime import datetime, timezone
+from typing import TYPE_CHECKING, Annotated, Any, Dict, Optional, Type
+
+from ape.logging import logger
+from pydantic import BaseModel
+from taskiq import TaskiqResult
+from typing_extensions import Self  # Introduced 3.11
+
+from .types import SilverbackIdent
+
+
+class SilverbackState(BaseModel):
+    instance: str
+    network: str
+    # Last block number seen by runner
+    last_block_seen: int
+    # Last block number processed by a worker
+    last_block_processed: int
+    updated: datetime
+
+
+class HandlerResult(BaseModel):
+    instance: str
+    network: str
+    handler_id: str
+    block_number: Optional[int]
+    log_index: Optional[int]
+    execution_time: float
+    # TODO: upcoming feature in taskiq
+    # labels: Dict[str]
+    # TODO: Use computed field with pydantic v2
+    return_value_blob: Optional[bytes]  # pickled data
+    created: datetime
+
+    @classmethod
+    def from_taskiq(
+        cls,
+        ident: SilverbackIdent,
+        handler_id: str,
+        block_number: int | None,
+        log_index: int | None,
+        result: TaskiqResult,
+    ) -> Self:
+        return cls(
+            instance=ident.identifier,
+            network=ident.network_choice,
+            handler_id=handler_id,
+            block_number=block_number,
+            log_index=log_index,
+            execution_time=result.execution_time,
+            # labels=result.labels,
+            return_value_blob=pickle.dumps(result.return_value),
+            created=datetime.now(timezone.utc),
+        )
+
+
+class BasePersistentStorage(ABC):
+    @abstractmethod
+    async def get_instance_state(self, ident: SilverbackIdent) -> Optional[SilverbackState]:
+        ...
+
+    @abstractmethod
+    async def set_instance_state(
+        self, ident: SilverbackIdent, last_block_seen: int, last_block_processed: int
+    ) -> Optional[SilverbackState]:
+        ...
+
+    @abstractmethod
+    async def get_latest_result(
+        self, instance: SilverbackIdent, handler: Optional[str] = None
+    ) -> Optional[HandlerResult]:
+        ...
+
+    @abstractmethod
+    async def add_result(self, v: HandlerResult):
+        ...
+
+
+async def init_mongo(mongo_uri: str) -> Optional[BasePersistentStorage]:
```

Comment:
There a better way to do this for optional dependencies?

Reply:
So pocket this for use elsewhere and implement SQLite?  I'm good with that, but a little worried about it scaling for anything real-world.

Reply:
alright, I tore out mongo and implemented SQLite.  Cleans things up a bit.  Implemented a better async init flow as well.

Reply:
> So pocket this for use elsewhere and implement SQLite?  I'm good with that, but a little worried about it scaling for anything real-world.

That's sorta the point. If you have a real-world use case for persistence, you already have a DB structure you intend to integrate with.

This should be more like some sort of Middleware that will migrate results as they come in from a different location, so if we give a way to hook into that process then anyone can do that however they feel the need to (postgres, mongo, etc.)

TaskIQ might even have a better way of providing this via result backends that support persisting data past the completion of the task

Reply:
> TaskIQ might even have a better way of providing this via result backends that support persisting data past the completion of the task

So, I guess, this makes me worried about the work here in this PR.  I'd like to think this is about 90% done and be able to move on to integrating it into chaosnet soon.

I currently have no reason to believe that their result backends will work for us.  Do you know something I don't?  Or have an idea on how we should leverage it?